src/storage/src/source/mod.rs

```diff
@@ -974,94 +952,101 @@ where
-
-        let mut source_reader = if !active {
-            None
-        } else {
-            match S::new(
-                name.clone(),
+
+        let sync_activator = scope.sync_activator_for(&info.address[..]);
+        let base_metrics = base_metrics.clone();
+        let source_connector = source_connector.clone();
+        let mut source_reader = Box::pin(async_stream::stream! {
+            let mut timestamper = reclock::Timestamper::new(storage_metadata, as_of, now).await;
+
+            // TODO(petrosagg): What is the purpose of an partition offset being None?
+            let start_offsets = timestamper.partition_cursors().into_iter().map(|(p, o)| (p, Some(o))).collect();
+
+            let source_reader = S::new(
+                name,
                 id,
                 worker_id,
                 worker_count,
-                scope.sync_activator_for(&info.address[..]),
+                sync_activator,
                 source_connector.clone(),
                 aws_external_id.clone(),
-                restored_offsets,
+                start_offsets,
                 encoding,
-                base_metrics.clone(),
-            ) {
-                Ok(source_reader) => Some(source_reader.into_stream(timestamp_frequency)),
+                base_metrics,
+            );
+            let source_stream = match source_reader {
+                Ok(s) => s.into_stream(timestamp_frequency).fuse(),
                 Err(e) => {
                     error!("Failed to create source: {}", e);
-                    None
-                }
-            }
-        };
-
-        move |cap, durability_cap, output| {
-            // First check that the source was successfully created
-            let source_reader = match &mut source_reader {
-                Some(source_reader) => source_reader,
-                None => {
-                    return SourceStatus::Done;
+                    return;
                 }
             };
 
-            // Check that we have a valid list of timestamp bindings.
-            let timestamp_histories = match &mut timestamp_histories {
-                Some(histories) => histories,
-                None => {
-                    error!("Source missing list of timestamp bindings");
-                    return SourceStatus::Done;
+            tokio::pin!(source_stream);
+
+            let mut timestamp_interval = tokio::time::interval(timestamp_frequency);
+            let mut pending_messages = vec![];
+            loop {
+                tokio::select! {
+                    item = source_stream.next() => {
+                        match item {
+                            Some(Ok(message)) => pending_messages.push(message),
+                            Some(Err(_)) => {
+                                panic!("Errors have no offset, and so cannot be remapped");
+                            }
+                            None => {},
+                        }
+                    }
```

Comment:
this implicitly requires source `next` impls to be cancel safe right? I am not sure if we can guarantee that, or if we rewrite this to avoid it: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=695213f1d001ad45e01f003e05940004, unless we somehow `spawn` the `next` call and wait-and-replace the `JoinHandle`, which will require we move the `SourceReader` in and our of threads on every message

Reply:
separately, we can just spawn a task and use a channel to move data between tokio and timely here, but it will be yet another move on what runtime work happens on

Reply:
I looked at the implementation and the only state held by the future is the reference to the stream and polling it just calls poll_next so it's cancel safe. Does this sound right?

Reply:
This `next` is the `SourceReader::into_stream`'s stream `next` right? that next calls `SourceReader::next`, whose default implementation is cancel-safe (well, technically you can skip over a transient delay delay or a Pending delay, but thats not a correctness problem), but people can override that impl, like `pubnub`, whose `next` is only cancel-safe if all the `pubnub-hyper` methods are, which I have no idea about

Reply:
It's StreamExt::next, no?

Reply:
Ultimately, that ends up calling the `SourceReader::next` tho

Reply:
Sure, but that is an implementation detail inside `async_stream::stream!` which is never dropped, we just call its `poll_next` method through the `.next()` call

Reply:
Ok, @petrosagg and I talked, because the `Next<'stream, Stream>` future only BORROWS the underlying stream, and the stream ITSELF is not dropped, this is actually safe

Reply:
If this is the code that is merged, then we should add a comment explaining this subtlety!